subsys/shell/backends/shell_uart.c

```diff
@@ -458,22 +480,27 @@ static int async_read(struct shell_uart_async *sh_uart,
 	bool buf_available = uart_async_rx_data_consume(async_rx, sh_cnt);
 	*cnt = sh_cnt;
 
-	if (sh_uart->pending_rx_req && buf_available) {
+	if (buf_available && (!sh_uart->rx_enabled || sh_uart->pending_rx_req)) {
 		uint8_t *buf = uart_async_rx_buf_req(async_rx);
 		size_t len = uart_async_rx_get_buf_len(async_rx);
 		int err;
 
 		__ASSERT_NO_MSG(buf != NULL);
 		atomic_dec(&sh_uart->pending_rx_req);
-		err = uart_rx_buf_rsp(sh_uart->common.dev, buf, len);
-		/* If it is too late and RX is disabled then re-enable it. */
-		if (err < 0) {
-			if (err == -EACCES) {
+		if (sh_uart->rx_enabled) {
+			err = uart_rx_buf_rsp(sh_uart->common.dev, buf, len);
+			/* If it is too late and RX is disabled then re-enable it. */
+			if (err < 0) {
+				/* Release allocated buffer. UART will be re-enabled
+				 * when UART_RX_DISABLED event is received.
+				 */
+				uart_async_rx_on_rdy(async_rx, buf, 0);
+				uart_async_rx_on_buf_rel(async_rx, buf);
 				sh_uart->pending_rx_req = 0;
-				err = rx_enable(sh_uart->common.dev, buf, len);
-			} else {
-				return err;
+				err = (err == -EACCES) ? 0 : err;
 			}
+		} else {
+			err = rx_enable(sh_uart->common.dev, sh_uart, buf, len);
 		}
```

Comment:
Some suggestions:
- we can probably refactor the `!sh_uart->rx_enabled` path (the `else` part) out and simply do `rx_enable` and return
  - looks like we are not carrying forward the return value of `rx_enable()` (it didn't before the change either), is that intended?
- then we do not need to nest the `sh_uart->rx_enabled` path